File: Golang/797. All Paths From Source to Target.go
// func allPathsSourceTarget(graph [][]int) [][]int { // 广度搜索, 运行不了
//     target := len(graph) - 1
//     currlevel := [][][]int{}
//     currlevel[0] = [][]int{}
//     currlevel[0][0] = []int{0}
//     currlevel[0][1] = []int{0}
//     ans := make([][]int, 0, len(graph))
    
//     for len(currlevel) > 0 {
//         nextlevel := [][][]int{}
//         for len(currlevel) > 0 {
//             curr := currlevel[len(currlevel)-1][0][0]
//             nodes := currlevel[len(currlevel)-1][1]
//             currlevel = currlevel[: len(currlevel)-1]
//             if curr == target {
//                 ans = append(ans, nodes)
//             }
//             for _, newnode := range graph[curr] {
//                 item := [][]int{}
//                 item[0][0] = newnode
//                 item[1] = append(nodes, newnode)
//                 nextlevel = append(nextlevel, item)
//             }
//         }
//         currlevel = nextlevel
//     }
//     return ans
// }

// func allPathsSourceTarget(graph [][]int) [][]int { // 递归, 依然报错很奇怪
//     ans := [][]int{}
//     target := len(graph) - 1
//     recursion(&graph, &ans, 0, target, []int{0})
//     return ans
// }

// func recursion(graph, ans *[][]int, i, target int, path []int) {
//     if i == target {
//         *ans = append(*ans, path)
//         return
//     }
//     for _, j := range (*graph)[i] {
//         recursion(graph, ans, j, target, append(path, j))
//     }
// }

func allPathsSourceTarget(graph [][]int) [][]int { // 深度搜索 还是没有结果
    target := len(graph) - 1
    currlevel := make([][][]int, 0, len(graph))
    currlevel[0] = make([][]int, 0, 2)
    currlevel[0][0] = []int{0}
    currlevel[0][1] = []int{0}
    ans := make([][]int, 0, len(graph))
    
    for len(currlevel) > 0 {

        curr := currlevel[len(currlevel)-1][0][0]
        nodes := currlevel[len(currlevel)-1][1]
        currlevel = currlevel[: len(currlevel)-1]
        if curr == target {
            ans = append(ans, nodes)
        }
        for _, newnode := range graph[curr] {
            item := [][]int{}
            item[0][0] = newnode
            item[1] = append(nodes, newnode)
            currlevel = append(currlevel, item)
        }

    }
    return ans
}

File: Python3/797. All Paths From Source to Target.py
class Solution:
    def allPathsSourceTarget(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[List[int]]
        """
        # target = len(graph) - 1 # 广度搜索 172ms
        # currlevel = [(0, [0])]
        # ans = []
        # while currlevel:
        #     nextlevel = []
        #     while currlevel:
        #         curr, nodes = currlevel.pop()
        #         if curr == target:
        #             ans.append(nodes)
        #         for newnode in graph[curr]:
        #             nextlevel.append((newnode, nodes+[newnode]))
        #     currlevel = nextlevel
        # return ans
            
        # target = len(graph) - 1 # 深度搜索 160ms
        # stack = [(0, [0])]
        # ans = []
        # while stack:
        #     curr, nodes = stack.pop()
        #     if curr == target:
        #         ans.append(nodes)
        #     for newnode in graph[curr]:
        #         stack.append((newnode, nodes+[newnode]))
        # return ans
        
        # ans, target = [], len(graph) - 1 # 递归 168ms
        # def recursion(curr, nodes):
        #     if curr == target:
        #         ans.append(nodes)
        #         return
        #     for newnode in graph[curr]:
        #         recursion(newnode, nodes+[newnode])
        # recursion(0, [0])
        # return ans
                
        ans, target = [], len(graph) - 1 # 参考discuss中的方法, 和上面的写法一样, 但是只需要148ms
        def recursion(i, path):
            if i == LAST:
                ans.append(path)
                return
            for j in graph[i]:
                recursion(j, path + (j,))
        recursion(0, (0, ))
        return ans
